django/db/models/expressions.py

```diff
@@ -490,6 +490,27 @@ def desc(self):
         return OrderBy(self, descending=True)
 
 
+class ResolvedOuterRef(F):
+    """
+    An object that contains a reference to an outer query.
+
+    In this case, the reference to the outer query has been resolved, because the
+    inner query has been used as a subquery.
+    """
+    def as_sql(self, *args, **kwargs):
+        raise ValueError('This queryset contains a reference to an outer query, and cannot be '
+                         "evaluated in its own right")
+
+
+class OuterRef(F):
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        return ResolvedOuterRef(self.name)
+
+    def as_sql(self, *args, **kwargs):
+        raise ValueError('This queryset contains an unresolved reference to an outer query, and '
+                         'may not be evaluated.')
+
```

Comment:
@akaariai I needed to have a two-step, as otherwise we don't know if it's the _initial_ resolve_expression, which doesn't really do anything, or the _real_ one, that actually resolves.